File: LeetCode/0314.Binary-Tree-Vertical-Order-Traversal/Binary-Tree-Vertical-Order-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        int min = 0;
        int max = 0;
        Map<Integer, ArrayList<Integer>> map = new HashMap<>();
        Queue<TreeNode> queue = new LinkedList<>();
        Queue<Integer> col = new LinkedList<>();
        List<List<Integer>> res = new ArrayList<>();
        if(root == null) return res;
        queue.offer(root);
        col.offer(0);
        while(!queue.isEmpty()){
            TreeNode cur = queue.poll();
            int colNum = col.poll();
            if(!map.containsKey(colNum)){
                map.put(colNum, new ArrayList<>());
            }
            map.get(colNum).add(cur.val);
            if(cur.left != null){
                queue.offer(cur.left);
                col.offer(colNum - 1);
                min = Math.min(min, colNum - 1);
            }
            if(cur.right != null){
                queue.offer(cur.right);
                col.offer(colNum + 1);
                max = Math.max(max, colNum + 1);
            }
        }
        for(int i = min; i <= max; i++){
            res.add(map.get(i));
        }
        return res;
    }
}

File: LeetCode/0314.Binary-Tree-Vertical-Order-Traversal/Binary-Tree-Vertical-Order-Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Solution(object):
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root: return []
        queue = deque()
        queue.append((root, 0, 0))
        res_r = []
        res_l = []
        while queue:
            node, row, col = queue.popleft()
            if col >= 0:
                for _ in range(col - len(res_r) + 1): res_r.append([])
                res_r[col].append(node.val)
            else:
                for _ in range(-col - len(res_l) + 1): res_l.append([])    
                res_l[-col].append(node.val)
            #print(node.val, row, col, res_r, res_l)
            if node.left:
                queue.append((node.left, row + 1, col - 1))
            if node.right:
                queue.append((node.right, row + 1, col + 1))
        res = []
        for i in range(len(res_l) - 1, -1, -1):
            if res_l[i] != []: res.append(res_l[i])
        for i in range(len(res_r)):
            if res_r[i] != []: res.append(res_r[i])
        return res
